### ISSUE TITLE
Kotlin Codegen Release Blockers

### ISSUE THREAD

**Issue by swankjesse** [square/moshi#461]
The code’s checked in but there’s some follow-ups to do before we can invite users to use it:

 - [x] kotlin-metadata dependency released in mavencentral (https://github.com/Takhion/kotlin-metadata/issues/5)
 - [x] constructor signature
 - [x] @JsonQualifier annotations
 - [x] @MoshiSerializable annotation and factory potentially in core
 - [x] ”Generated code; do not edit“ header

**Comment by ChrisAckerman** [comment]
Add jcenter to your pom, and there’s no need for kotlin-metadata to be in maven central.

```
<repositories>
    <repository>
      <id>jcenter</id>
      <url>https://jcenter.bintray.com/</url>
    </repository>
</repositories>
```

**Comment by ChrisAckerman** [comment]
https://bintray.com/takhion/maven/kotlin-metadata

**Comment by JakeWharton** [comment]
Sonatype OSSRH prevents release of artifacts to Maven central which point to arbitrary repositories.

**Comment by hzsweers** [comment]
We should probably generated `@Generated` annotations onto the adapters via https://github.com/google/auto/blob/master/common/src/main/java/com/google/auto/common/GeneratedAnnotationSpecs.java#L46-L61

This could possibly just replace the "generated code, do not edit" header bit in the list. 

This would also allow for more fine-grained proguard `keepnames` retentions, which otherwise would have to keep all of moshi's built-in `JsonAdapter`-suffixed classes as well due to `**JsonAdapter`

```proguard
-keepnames @com.squareup.moshi.<wherever this ends up>.MoshiSerializable class *

# Java < 9
-keepnames @javax.annotation.Generated class **JsonAdapter

#  Java 9+
-keepnames @javax.annotation.processing.Generated class **JsonAdapter
```

**Comment by hzsweers** [comment]
Another one for the list: proguard will make seemingly unused constructors `private`, so we should defensively check in `MoshiSerializableFactory` if the `Constructor` is accessible and make it so if it's not

**Comment by swankjesse** [comment]
I’m looking into this. We need something like compile-testing for Kotlin. Otherwise we can’t test that unsupported inputs like abstract classes are handled appropriately. Right now we’re only testing inputs that can successfully generate an adapter.

**Comment by hzsweers** [comment]
@takhion has actually mentioned working on something like that recently,
not sure where it stands.




On Mar 28, 2018 at 7:39 PM, Jesse Wilson <notifications@github.com> wrote:


I’m looking into this. We need something like compile-testing for Kotlin.
Otherwise we can’t test that unsupported inputs like abstract classes are
handled appropriately. Right now we’re only testing inputs that can
successfully generate an adapter.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/square/moshi/issues/461#issuecomment-377102128>, or mute
the thread

**Comment by Takhion** [comment]
@swankjesse @hzsweers

### `kotlin-meadata` in Maven Central
Everything is ready, just waiting on [OSSRH-38743](https://issues.sonatype.org/browse/OSSRH-38743).

### `compile-testing` for Kotlin
I'm developing a prototype for `kotlin-metadata` based on [Gradle TestKit](https://docs.gradle.org/current/userguide/test_kit.html), but it's gonna take some more time!

**Comment by LouisCAD** [comment]
@bishiboosh Kotshi can already save your life: https://github.com/ansman/kotshi

**Comment by bishiboosh** [comment]
I know about kotshi, but I would rather use an officially sanctionned implementation (plus this one handles default values better, and will support qualifiers properly)

**Comment by LouisCAD** [comment]
@bishiboosh FYI, Jake Wharton committed on Kotshi repo. Also, this is open source stuff, so no need for an "official" thing approved by a president 😆. Yes, default values are better handled there, but if you don't have too much of them, using Kotshi until this gets released might be a good solution for you.

**Comment by LouisCAD** [comment]
@Takhion It seems that they are waiting for a comment response [there](https://issues.sonatype.org/browse/OSSRH-38743) in order to host your project on mavenCentral.

**Comment by Takhion** [comment]
@LouisCAD DOH! Thank you, I assumed that was just an automatic reply (which it is), but didn't see the `Waiting for Response` status 🤦‍♂️! Just [replied](https://issues.sonatype.org/browse/OSSRH-38743?focusedCommentId=471727&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-471727), let's see...

**Comment by LouisCAD** [comment]
@Takhion Your comment response definitely changed the status! Back to open now 😃

**Comment by bishiboosh** [comment]
Kind-of stupid question, but as this is a bug in Kotshi, it may be useful here too : are `@Transient` field handled properly when creating the adapter ? (as in, ignored when serializing and set to default value when deserializing)

**Comment by LouisCAD** [comment]
@swankjesse The checkboxes are out of date, `kotlin-metadata` is now on mavenCentral

**Comment by swankjesse** [comment]
I’d like to drop support for types that are subtypes of concrete types. The problem is that when we codegen a class we will need to observe the properties of its superclass. But the superclass could change (maybe it’s in a different project!) and we won’t regenerate the subclass’ adapter.

This is easy with reflection (it’s late binding) but nearly impossible with codegen.

Thoughts?

**Comment by LouisCAD** [comment]
@swankjesse Could you illustrate what you're saying with example code snippets?

**Comment by Takhion** [comment]
@swankjesse the annotation processor always runs (unless incremental processing is explicitly enabled) so changes to any class should always be tracked, or am I missing something?

**Comment by NightlyNexus** [comment]
@Takhion that seems reliant on a build system. also, somebody might distribute generated sources and not the generator.

>drop support for types that are subtypes of concrete types.
...
we will need to observe the properties of its superclass. But the superclass could change (maybe it’s in a different project!) and we won’t regenerate the subclass’ adapter.

Doesn't that argument apply to types that are subtypes of **abstract** types, too?

I'm a fan of killing off inheritance to avoid accidents. On one hand, it'd be nice to have parity with the reflective adapters. On the other, I think the annotation processor can be be more opinionated. It's a power user feature.

**Comment by Takhion** [comment]
@NightlyNexus but that's true of every annotation processor and pretty much expected behaviour: in general you should only use it at the "final" stage of compilation (like when producing the consumer executable/app), unless you know _exactly_ what you're doing :)

**Comment by JakeWharton** [comment]
Dagger and Butter Knife are two counterpoints to that.

**Comment by NightlyNexus** [comment]
yeah, i immediately thought of Dagger.

**Comment by swankjesse** [comment]
It’s done. Yay!
